src/library/scala/collection/immutable/OrderedMap.scala

```diff
@@ -0,0 +1,426 @@
+
+    override def hasNext: Boolean = iter.hasNext
+
+    override def next(): K = iter.next()
+  }
+
+  override def valuesIterator: Iterator[V] = new AbstractIterator[V] {
+    private[this] val iter = ordering.iterator
+
+    override def hasNext: Boolean = iter.hasNext
+
+    override def next(): V = value(binding(iter.next()))
+  }
+
+  override def concat[V1 >: V](suffix: IterableOnce[(K, V1)]): OrderedMap[K, V1] = {
+    var result: OrderedMap[K, V1] = this
+    val iter = suffix.iterator
+    while (iter.hasNext) {
+      val (k, v) = iter.next()
+      result = result.updated(k, v)
+    }
+    result
+  }
+
+  override def map[K2, V2](f: ((K, V)) => (K2, V2)): OrderedMap[K2, V2] = {
+    var ong: Ordering[K2] = Ordering.empty
+    val mng: Mapping[K2, V2] = mapping.map {
+      case (k, (o, v)) =>
+        val (k2, v2) = f((k, v))
+        ong = ong.incl(o, k2)
+        (k2, (o, v2))
+    }
+    new OrderedMap[K2, V2](ong, mng, ordinal, orderedBy)
+  }
+
+  override def flatMap[K2, V2](f: ((K, V)) => IterableOnce[(K2, V2)]): OrderedMap[K2, V2] = {
+    var ong: Ordering[K2] = Ordering.empty
+    val mng: Mapping[K2, V2] = mapping.flatMap {
+      case (k, (o, v)) =>
+        f((k, v)).iterator.reduceLeftOption((_, e) => e).map {
+          case (k2, v2) =>
+            ong = ong.incl(o, k2)
+            (k2, (o, v2))
+        }
+    }
+    new OrderedMap[K2, V2](ong, mng, ordinal, orderedBy)
+  }
+
+  override def collect[K2, V2](pf: PartialFunction[(K, V), (K2, V2)]): OrderedMap[K2, V2] = {
+    var ong: Ordering[K2] = Ordering.empty
+    val mng: Mapping[K2, V2] = mapping.collect {
+      case (k, (o, v)) if pf.isDefinedAt((k, v)) =>
+        val (k2, v2) = pf((k, v))
+        ong = ong.incl(o, k2)
+        (k2, (o, v2))
+    }
+    new OrderedMap[K2, V2](ong, mng, ordinal, orderedBy)
+  }
+
+  @`inline` private[this] def value(p: (_, V)) = p._2
+  @`inline` private[this] def binding(k: K) = mapping(k).copy(_1 = k)
+}
+object OrderedMap extends MapFactory[OrderedMap] {
+  sealed trait OrderBy
+  final object OrderBy {
+    final case object Insertion extends OrderBy
+    final case object Modification extends OrderBy
+  }
+
+  val Empty = new OrderedMap[Nothing, Nothing](Ordering.empty, HashMap.empty, 0, OrderBy.Insertion)
+  def empty[K, V]: OrderedMap[K, V] = empty(OrderBy.Insertion)
+  def empty[K, V](orderBy: OrderBy): OrderedMap[K, V] = Empty.asInstanceOf[OrderedMap[K, V]]
+
+  def from[K, V](it: collection.IterableOnce[(K, V)]): OrderedMap[K, V] =
+    it match {
+      case vm: OrderedMap[K, V] => vm
+      case _                   => (newBuilder[K, V] ++= it).result()
+    }
+
+  def newBuilder[K, V]: mutable.Builder[(K, V), OrderedMap[K, V]] = newBuilder(OrderBy.Insertion)
+
+  def newBuilder[K, V](orderBy: OrderBy): mutable.Builder[(K, V), OrderedMap[K, V]] =
+    new mutable.ImmutableBuilder[(K, V), OrderedMap[K, V]](empty(orderBy)) {
+      def addOne(elem: (K, V)): this.type = { elems = elems + elem; this }
+    }
```

Comment:
Could create a mutating builder for much better performance.

Reply:
Same as above, I'll look at HashMap for guidance.

Reply:
Addressed by 00b7f1e.